### PR TITLE
Feature/namespaced model

### PR DESCRIPTION
**Pull request by saiqulhaq**
this is improved  PR for my previous PR
now model class like PurchasedOrders or Foo::Bar wouldn't be a problem, because we can set the model name manually

### CODE REVIEW

File: lib/ajax-datatables-rails/base.rb [hunk 1]
```diff
@@ -139,12 +154,17 @@ def per_page
     end
 
     def sort_column(item)
-      sortable_columns[item['column'].to_i]
```
Comment (saiqulhaq) [inline lib/ajax-datatables-rails/base.rb:142]: because the model name customized, so the sort column should be customized too

**Comment by antillas21** [comment]
@saiqulhaq 

After reviewing your proposed changes, which have helped me realize some shortcomings of the original design (thank you for that :smile: ).. I think the easier thing to do, is to change the way that `sortable_columns` and `searchable_columns` are declared. Perhaps the following changes are in the long run, easier for everyone else to understand:

``` ruby
def searchable_columns
  # declare strings in this format: ModelName.column_name
  @searchable_columns ||= %w(
      User.first_name User.last_name PurchaseOrder.number PurchaseOrder.created_at
      PurchaseOrder.total Purchase::LineItem.quantity Purchase::LineItem.unit_cost
      Purchase::LineItem.tax
    )
end

def sortable_columns
  # declare strings in this format: ModelName.column_name
  @sortable_columns ||= %w(
      User.first_name User.last_name PurchaseOrder.number PurchaseOrder.created_at
      PurchaseOrder.total Purchase::LineItem.quantity Purchase::LineItem.unit_cost
      Purchase::LineItem.tax
    )
end

def search_condition(column, value)
  model, column = column.split('.')
  model = model.constantize
  casted_column = ::Arel::Nodes::NamedFunction.new('CAST', [model.arel_table[column.to_sym].as(typecast)])
  casted_column.matches("%#{value}%")
end

def sort_column(item)
  model, column = sortable_columns[item[:column].to_i].split('.')
  col = [model.constantize.table_name, column].join('.')
end
```

Again... I pretty much like the train of thought you used to solve this feature... and most of all, I love the  time dedicated to fixing the spec suite :+1: .

Your approach on this helped me come up with a schema that relies less on other classes and that (I think) can feel "more declarative" and less confusing when defining what columns are available to search on and sort by.

What do you think?

**Comment by saiqulhaq** [comment]
actually, the main problem is how to get model name
i think your approach its better, because its not depend on safe_constantize like mine, so it wouldn't break spec on rails 3.1.1

**Comment by saiqulhaq** [comment]
but, 

``` ruby
model = model.constantize
```

is not compatible with previous code, so it will break current and previous spec 

so how about we passing parameter like

``` ruby
Foo::AjaxDatatablesRails.new view_context, namespaced: true
```

then 

``` ruby
def get_model_class_for(model)
  return model.constantize if @options.namespaced

  model.singularize.titleize.gsub( / /, '' ).constantize
end

def search_condition(column, value)
  model, column = column.split('.')
  model = get_model_class_for(model)
  casted_column = ::Arel::Nodes::NamedFunction.new('CAST', [model.arel_table[column.to_sym].as(typecast)])
  casted_column.matches("%#{value}%")
end

def sort_column(item)
  model, column = sortable_columns[item[:column].to_i].split('.')
  col = [get_model_class_for(model).table_name, column].join('.')
end
```

**Comment by antillas21** [comment]
Wouldn't it be better to leave namespacing (discovery) to the developer? after all, each developer would know the right way to their defined classes. Example:

``` ruby
def searchable_columns
  # declare strings in this format: ModelName.column_name
  @searchable_columns ||= %w(
      ::User.first_name ::User.last_name PurchaseOrder.number PurchaseOrder.created_at
      PurchaseOrder.total Purchase::LineItem.quantity Purchase::LineItem.unit_cost
      Purchase::LineItem.tax
    )
end
```

Notice how some models use `::` to make classes search start at the root namespace.

**Comment by antillas21** [comment]
My later comment is because, I wouldn't like to make a `Base` class depend on user provided params as options.

My original idea to provide users with an `options` hash, was to to allow the user to define custom methods and/or define how to retrieve data based on the options params received, because:
- user defines options
- options are used on user _defined_ methods

**Comment by saiqulhaq** [comment]
yeah i know that `namespaced: true` options it's not suitable
but how do you make compatibilty with current and previous version?

because currently, users using singularize name like `["users.name", "articles.title"]`, not `"user.name"` 
so it will break if just using `model.constantize`

**Comment by antillas21** [comment]
I think it can be supported for the immediate version with a deprecation message... Next version after that, support to use the previous version for declaring `searchable_columns` and `sortable_columns` shall be removed.

**Comment by antillas21** [comment]
@saiqulhaq please take a look at this branch: [version 0.3.0](https://github.com/antillas21/ajax-datatables-rails/tree/version-0-3-0)... this is where I merged in your contribution (for the time being, while I think a little more about a definitive change in the API).

You will notice that this branch supports both the (now) previous version for declaring `searchable_columns` and `sortable_columns` methods, as well as the changes we are discussing here.

**Comment by saiqulhaq** [comment]
i have been read the v0.3.0 branch, and i really like the `#search_condition` implementation :+1: 

hmm so, you wanna remove `searchable_columns` and `sortable_columns` methods at next version?

**Comment by antillas21** [comment]
No, I'm not thinking about removing these methods... the plan is:
- support both ways to define `searchable_columns`, `sortable_columns` (who by the time being use `new_` and `deprecated_` prefixes on internal methods, depending on how the user declares them).
- next gem release (0.4.0) will drop support for all `deprecated_` prefixed methods.
- we will continue to rely on `searchable_columns` and `sortable_columns` methods to define how to search and sort records. But these will have to be defined using the new syntax starting on the next release after 0.3.0

**Comment by saiqulhaq** [comment]
ok, i'm understand
so the plan is to change users syntax from "models.foo" to Model.foo

**Comment by antillas21** [comment]
Yes... starting on this upcoming version 0.3.0 (with backwards compatibility), and from version 0.4.0 forward no backwards compatibility.

This is small plan for version 0.3.0: https://github.com/antillas21/ajax-datatables-rails/milestones/v%200.3.0

**Comment by antillas21** [comment]
@saiqulhaq I am closing the PR, for administrative purposes.

**Please note** that this PR has been accepted and merged manually and is scheduled to be included in the next version release (0.3.0) once all the small features that compose that version are done.

I'd like to thank you for your contribution and for pointing out shortcomings in our original API design. :+1:

**Comment by saiqulhaq** [comment]
you're welcome

**Pull request closed without merging**